account-compression/programs/account-compression/src/lib.rs

```diff
@@ -173,69 +212,158 @@ pub mod spl_account_compression {
+                return Err(AccountCompressionError::IncorrectAccountOwner.into());
+            }
+            let space = (max_depth * 32) + 40;
+            let lamports = Rent::get()?.minimum_balance(space as usize);
+            let create_ix = create_account(
+                &ctx.accounts.payer.key,
+                &ctx.accounts.proof_buffer.key,
+                lamports,
+                space as u64,
+                &crate::id(),
+            );
+            invoke(
+                &create_ix,
+                &[
+                    ctx.accounts.payer.to_account_info(),
+                    ctx.accounts.proof_buffer.to_account_info(),
+                    ctx.accounts.system_program.to_account_info(),
+                ],
+            )?;
+            ProofBufferHeader {
+                owner: ctx.accounts.payer.key.key(),
+                remaining_depth: max_depth,
+                max_depth,
+            }
+        } else {
+            let (header, _) = buffer.split_at_mut(40);
+            *ProofBufferHeader::load_mut_bytes(header)?
+        };
+        if ctx.accounts.payer.key.as_ref() != header.owner.as_ref() {
+            return Err(AccountCompressionError::InvalidProofBufferOwner.into());
+        }
+        if header.max_depth != max_depth {
+            return Err(AccountCompressionError::ProofIndexOutOfBounds.into());
+        }
+        let offset = 40 + (index * 32) as usize;
+        let segments: Vec<&[u8]> = partial_proof.chunks(32).collect();
+        for proof in segments.iter() {
+            buffer[offset..40 + ((offset + 1) * 32)].copy_from_slice(proof);
+        }
+        let new_remaining = header.remaining_depth - index;
+        //set remaining
+        header.remaining_depth = new_remaining;
+        Ok(())
+    }
+
+    /// Closes the proof buffer account
+    pub fn close_proof_buffer(ctx: Context<CloseProofBuffer>) -> Result<()> {
+        let data = ctx.accounts.proof_buffer.try_borrow_data()?;
+        let (owner, _) = data.split_at(32);
+        if owner != ctx.accounts.payer.key.as_ref() {
+            return Err(AccountCompressionError::InvalidProofBufferOwner.into());
+        }
+        let proof_lamports = ctx.accounts.proof_buffer.get_lamports();
+        ctx.accounts.proof_buffer.sub_lamports(proof_lamports)?;
+        ctx.accounts.payer.add_lamports(proof_lamports)?;
+        Ok(())
+    }
+
+    // Creates a new merkle tree with an existing root in the buffer.
+    // The indexers for the specific tree use case are in charge of handling the indexing or shadow indexing of the tree.
+    // Shadow indexing is a technique that allows for the indexing of a tree incrementally with update operations.
+    pub fn init_merkle_tree_with_root(
+        ctx: Context<InitializeWithRoot>,
+        max_depth: u32,
+        max_buffer_size: u32,
+        root: [u8; 32],
+        leaf_index: u32,
+        leaf: [u8; 32], // the last leaf in the tree
+        manifest_url: String,
```

Comment:
Great work, @austbot, this should unlock projects in rolling out more compression. As for the manifest_url,  wouldn't it be more convenient to leave this logic to the caller of the `init_merkle_tree_with_root` like the bubblegum contract? As devs may want to enreach the relevent events with other information for the indexers as well, wdyt?